File: crates/breez-sdk/bindings/examples/cli/langs/swift/Sources/BreezCLI/Contacts.swift
import Foundation
import BreezSdkSpark

// MARK: - Contacts command names (for REPL completion)

let contactsCommandNames: [String] = [
    "contacts add",
    "contacts update",
    "contacts delete",
    "contacts list",
]

// MARK: - Dispatch

func dispatchContactsCommand(_ args: [String], sdk: BreezSdk) async {
    if args.isEmpty || args[0] == "help" {
        printContactsHelp()
        return
    }

    let subName = args[0]
    let subArgs = Array(args.dropFirst())

    do {
        switch subName {
        case "add":
            try await handleContactAdd(sdk, subArgs)
        case "update":
            try await handleContactUpdate(sdk, subArgs)
        case "delete":
            try await handleContactDelete(sdk, subArgs)
        case "list":
            try await handleContactList(sdk, subArgs)
        default:
            print("Unknown contacts subcommand: \(subName). Use 'contacts help' for available commands.")
        }
    } catch {
        print("Error: \(error)")
    }
}

// MARK: - Help

private func printContactsHelp() {
    print("\nContacts subcommands:")
    print("  contacts \("add".padding(toLength: 30, withPad: " ", startingAt: 0))Add a new contact")
    print("  contacts \("update".padding(toLength: 30, withPad: " ", startingAt: 0))Update an existing contact")
    print("  contacts \("delete".padding(toLength: 30, withPad: " ", startingAt: 0))Delete a contact")
    print("  contacts \("list".padding(toLength: 30, withPad: " ", startingAt: 0))List contacts")
    print()
}

// MARK: - Handlers

// --- add ---

private func handleContactAdd(_ sdk: BreezSdk, _ args: [String]) async throws {
    let fp = FlagParser(args)
    guard fp.positional.count >= 2 else {
        print("Usage: contacts add <name> <payment_identifier>")
        return
    }

    let result = try await sdk.addContact(
        request: AddContactRequest(
            name: fp.positional[0],
            paymentIdentifier: fp.positional[1]
        ))
    printValue(result)
}

// --- update ---

private func handleContactUpdate(_ sdk: BreezSdk, _ args: [String]) async throws {
    let fp = FlagParser(args)
    guard fp.positional.count >= 3 else {
        print("Usage: contacts update <id> <name> <payment_identifier>")
        return
    }

    let result = try await sdk.updateContact(
        request: UpdateContactRequest(
            id: fp.positional[0],
            name: fp.positional[1],
            paymentIdentifier: fp.positional[2]
        ))
    printValue(result)
}

// --- delete ---

private func handleContactDelete(_ sdk: BreezSdk, _ args: [String]) async throws {
    guard !args.isEmpty else {
        print("Usage: contacts delete <id>")
        return
    }

    try await sdk.deleteContact(id: args[0])
    print("Contact deleted successfully")
}

// --- list ---

private func handleContactList(_ sdk: BreezSdk, _ args: [String]) async throws {
    let fp = FlagParser(args)
    let offset = fp.get("offset").flatMap { UInt32($0) }
        ?? fp.positional.first.flatMap { UInt32($0) }
    let limit = fp.get("limit").flatMap { UInt32($0) }
        ?? (fp.positional.count > 1 ? UInt32(fp.positional[1]) : nil)

    let result = try await sdk.listContacts(
        request: ListContactsRequest(
            offset: offset,
            limit: limit
        ))
    printValue(result)
}

File: crates/breez-sdk/bindings/examples/cli/langs/kotlin-multiplatform/src/main/kotlin/Contacts.kt
import breez_sdk_spark.*
import org.jline.reader.LineReader

/**
 * Represents a single contacts subcommand.
 */
data class ContactsCliCommand(
    val name: String,
    val description: String,
    val run: suspend (sdk: BreezSdk, reader: LineReader, args: List<String>) -> Unit
)

/**
 * All contacts subcommand names.
 */
val CONTACTS_COMMAND_NAMES = listOf(
    "add",
    "update",
    "delete",
    "list",
)

/**
 * Builds the contacts command registry.
 */
fun buildContactsRegistry(): Map<String, ContactsCliCommand> {
    return mapOf(
        "add" to ContactsCliCommand("add", "Add a new contact", ::handleAddContact),
        "update" to ContactsCliCommand("update", "Update an existing contact", ::handleUpdateContact),
        "delete" to ContactsCliCommand("delete", "Delete a contact", ::handleDeleteContact),
        "list" to ContactsCliCommand("list", "List contacts", ::handleListContacts),
    )
}

/**
 * Dispatches a contacts subcommand.
 */
suspend fun dispatchContactsCommand(
    args: List<String>,
    sdk: BreezSdk,
    registry: Map<String, ContactsCliCommand>,
    reader: LineReader,
) {
    if (args.isEmpty() || args[0] == "help") {
        println()
        println("Contacts subcommands:")
        registry.keys.sorted().forEach { name ->
            val cmd = registry[name]!!
            println("  contacts %-26s %s".format(name, cmd.description))
        }
        println()
        return
    }

    val subName = args[0]
    val subArgs = args.drop(1)

    val cmd = registry[subName]
    if (cmd == null) {
        println("Unknown contacts subcommand: $subName. Use 'contacts help' for available commands.")
        return
    }

    try {
        cmd.run(sdk, reader, subArgs)
    } catch (e: Exception) {
        println("Error: ${e.message}")
    }
}

// ---------------------------------------------------------------------------
// Contacts command handlers
// ---------------------------------------------------------------------------

// --- add ---

suspend fun handleAddContact(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    if (args.size < 2) {
        println("Usage: contacts add <name> <payment_identifier>")
        return
    }

    val contact = sdk.addContact(
        AddContactRequest(
            name = args[0],
            paymentIdentifier = args[1],
        )
    )
    printValue(contact)
}

// --- update ---

suspend fun handleUpdateContact(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    if (args.size < 3) {
        println("Usage: contacts update <id> <name> <payment_identifier>")
        return
    }

    val contact = sdk.updateContact(
        UpdateContactRequest(
            id = args[0],
            name = args[1],
            paymentIdentifier = args[2],
        )
    )
    printValue(contact)
}

// --- delete ---

suspend fun handleDeleteContact(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    if (args.isEmpty()) {
        println("Usage: contacts delete <id>")
        return
    }

    sdk.deleteContact(args[0])
    println("Contact deleted successfully")
}

// --- list ---

suspend fun handleListContacts(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    val fp = FlagParser(args)
    val offset = fp.getUInt("offset") ?: fp.positional.getOrNull(0)?.toUIntOrNull()
    val limit = fp.getUInt("limit") ?: fp.positional.getOrNull(1)?.toUIntOrNull()

    val contacts = sdk.listContacts(
        ListContactsRequest(
            offset = offset,
            limit = limit,
        )
    )
    printValue(contacts)
}
